src/pathfinding_igvc/src/PathFinding.cpp

```diff
@@ -0,0 +1,157 @@
+ * Path call back. Produces twist msg and publishes it
+ */
+void PathFinding::pathCallBack(const nav_msgs::Path::ConstPtr& path_ptr) {
+    nav_msgs::Path path_msg = *path_ptr; //Take required information from received message
+    geometry_msgs::Twist twist_msg = pathToTwist(path_msg);
+    twist_publisher.publish(twist_msg);
+}
+
+/**
+ * Stores the current position and orientation of the robot in the global frame
+ */
+void PathFinding::tfCallBack(const tf2_msgs::TFMessageConstPtr tf_message) {
+    // Check if the tf_message contains the transform we're looking for
+    for (geometry_msgs::TransformStamped tf_stamped : tf_message->transforms) {
+
+        //header frame might be base
+        if (tf_stamped.header.frame_id == global_frame && tf_stamped.child_frame_id == base_frame) {
+            // We've found the transform we're looking for, so see how close we
+            // are to the waypoint
+            double xPos = tf_stamped.transform.translation.x;
+            double yPos = tf_stamped.transform.translation.y;
+
+            double quatx = tf_stamped.transform.rotation.x;
+            double quaty = tf_stamped.transform.rotation.y;
+            double quatz = tf_stamped.transform.rotation.z;
+            double quatw = tf_stamped.transform.rotation.w;
+            tf::Quaternion q(quatx, quaty, quatz, quatw);
+            tf::Matrix3x3 m(q);
+            double roll, pitch, yaw;
+            m.getRPY(roll, pitch, yaw);
+
+            //Set member variables
+            robotXPos = xPos;
+            robotYPos = yPos;
+            robotOrientation = yaw; //Orientation = rotation about z axis
+        }
+    }
+}
+
+/**
+ * Processes path msg and produces corresponding twist msg
+ * @param path_msg
+ * @return a twist msg
+ */
+geometry_msgs::Twist PathFinding::pathToTwist(nav_msgs::Path path_msg) {
+    geometry_msgs::Twist twist_msg; //Initialize velocity message
+    const int NUM_VECTORS = 10; //Number of vectors to process including initial robot position
+
+    std::vector<geometry_msgs::PoseStamped> inc_poses = path_msg.poses;
+
+    //If we can assume poses are in order of priority then we don't need position?
+    //***May need curr position of robot as a starting "pose"
+    //Are poses frequently updated? Can we always assume the head of the poses array contains relevant data?
+    std::vector<float> x_vectors; //holds x values for vectors
+    std::vector<float> y_vectors; //holds y values for vectors
+    calcVectors(inc_poses, x_vectors, y_vectors, NUM_VECTORS); //x_vectors and y_vectors now updated
+
+    float x_sum = weightedXSum(x_vectors,NUM_VECTORS);
+    float y_sum = weightedYSum(y_vectors,NUM_VECTORS);
+
+    float desired_angle = atan(y_sum/x_sum);
+    float current_angle = robotOrientation;
+
+    float turn_rate = desired_angle - current_angle;
+    float speed = 1.0 - fabs(fmod(turn_rate,2*M_PI)/(2*M_PI));
+
+    twist_msg.linear.x = speed;
+    twist_msg.angular.z = turn_rate;
+
+    return twist_msg;
```

Comment:
In order to make testing easier, ideally we would have a single function `static` function that you can pass the current position/rotation of the robot and the path into, and which produces a  `Twist` message.